src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/WMSCascadeTest.java

```diff
@@ -189,13 +188,54 @@ public void testCascadeGetLegendRequestJSON() throws Exception {
 
         print(dom);
         JSONObject responseJson = JSONObject.fromObject(dom.toString());
-        assertTrue(responseJson.has(JSONLegendGraphicBuilder.LEGEND));
+        // styless remote layer..fixed wrong assertion
+        assertTrue(responseJson.isEmpty());
+    }
 
-        JSONArray legendArray = responseJson.getJSONArray(JSONLegendGraphicBuilder.LEGEND);
-        assertFalse(legendArray.isEmpty());
+    @Test
+    public void testScaleDenominator() throws Exception {
+        LayerGroupInfo info = getCatalog().getLayerGroupByName("cascaded_group");
+        LayerInfo groupLayer2 = getCatalog().getLayerByName("group_lyr_2");
+        assertNotNull(info);
+        assertNotNull(groupLayer2);
+        ReferencedEnvelope request1 =
+                new ReferencedEnvelope(groupLayer2.getResource().getNativeBoundingBox());
+        // minx,miny,maxx,maxy
+        String lyrBBox =
+                request1.getMinX()
+                        + ","
+                        + request1.getMinY()
+                        + ","
+                        + request1.getMaxX()
+                        + ","
+                        + request1.getMaxY();
+
+        // configure max scale on one of the layers
+        // set max scale as small as possible to have it filtered
+        WMSLayerInfo groupLayer1WMSResource =
+                (WMSLayerInfo) getCatalog().getLayerByName("group_lyr_1").getResource();
+        groupLayer1WMSResource.setMinScale(1);
+        groupLayer1WMSResource.setMaxScale(1000);
+        getCatalog().save(groupLayer1WMSResource);
 
-        JSONArray rulesJSONArray =
-                legendArray.getJSONObject(0).getJSONArray(JSONLegendGraphicBuilder.RULES);
-        assertFalse(rulesJSONArray.isEmpty());
+        String getMapRequest =
+                "wms?service=WMS&version=1.1.0"
+                        + "&request=GetMap"
+                        + "&layers="
+                        + info.getName()
+                        + "&bbox="
+                        + lyrBBox
+                        + "&width=768&height=537&srs=EPSG:4326&Format=image/png";
+
+        // should result in a request with both group layers present
+        // should invoke expected Mock URL in which both layers are present
+        // since the BBOX covers both
+        // but group_lyr_1 min/max scale is outside the bound of map scale
+        BufferedImage response = getAsImage(getMapRequest, "image/png");
```

Comment:
It's not really testing that the output is correct though. See ImageAssert in GeoTools to actually check the output.

Reply:
Allow me to explain, basically the test asserts that the GetMap request outside the scale range of one of the cascaded layer in the group used in integration test generates the expected URL (registered in MockHttpClient), which is the external WMS request should not include the layer configured with scale denominators. So if the integration request fails or generates a URL that was not expected (e.g layer is present even outside scale, etc), then this line will fail.